doc/design/fully_static_graph.md

```diff
@@ -0,0 +1,122 @@
+# Design Doc: Fully Static Graph
+
+## Abstract
+
+We propose the *fully static graph* rule: training and inference must
+be fully specified by the static graph. This means training and
+inference should be able to run solely on the cpp core (no Python
+involved), everything should be implemented as an OP.
+
+The user can still use Python to achieve the same result for
+convenience when experimenting locally, but the distributed training
+will not support Python.
+
+## Background
+
+There are two paradigms for expressing the computation graph: dynamic
+and static. The dynamic paradigm constructs the graph on the fly:
+every time `eval` is called, a new graph is created. The static
+paradigm constructs the graph first, and then calls `eval`. There is
+no new graph created each time `eval` is called.
+
+The dynamic graph has the advantage of being flexible but is highly
+dependent on the host language (most commonly Python). The static
+graph is not as flexible, but more optimization can be done since the
+graph is known before computing happens. PaddlePaddle is using the
```

Comment:
I agree with you that the fault recovery is easier to do if we always use a serializable computational graph and I also agree that the dist-paddle should never know `Python` and how that computational graph generated.

But I prefer that The `dist-paddle` just run that graph at once and do not care whether that `graph` contains a `for-loop` or not. It should be determined by the client-side. Because

1. It is clean and straightforward. The `dist-paddle` just run a graph. The user program just generates the graph and invoke `dist-paddle`.
2. It is flexible. The user can create a whole graph or make many graphs for one training process (think about a pre-train network and a real network run in the same training process).

If we implement our framework like that, we should make the user experience in local mode as same as in distributed mode. We should consider the following things

1. In local mode, the running logic should be same. Instead of sending the computation graph to multiple distributed Paddle, we just send it to local Paddle directly.
2. In local mode, we can `find` or `create` a variable in Python. For distributed Paddle, where does the user `find` a variable? Maybe a `distributed scope`?
3. Maybe we should not let the user define a `data reader` in Python. Users can only read their data by `data_reader` operator, which convert files into mini-batches. But there is a difference between local mode and distributed mode of `data_reader` operator is reading from local filesystem or reading from a distributed filesystem. And one more thing for the distributed `data_reader` operator is that the multiple nodes should read the whole epoch of data but each node should read the different part of them.

Reply:
Thanks for the very valuable suggestion! I will reconsider the solutions and discuss with you offline.

Reply:
Thanks! Updated.